File: src/02.string/04.transform/08.640.ts
/**
 * 3 <= equation.length <= 1000
 * equation 只有一个 '='.
 * 方程由绝对值在 [0, 100]  范围内且无任何前导零的整数和变量 'x' 组成
 * @param equation
 * @returns
 */

export function solveEquation(equation: string): string {
    let coefficient = 0;
    let constant = 0;
    let i = 0;
    let n = equation.length;
    let hasLeft = true;

    while (i < n) {
        if (equation[i] === "=") {
            i++;
            hasLeft = false;
            continue;
        }

        let curValue = hasLeft ? "" : "-";
        if (equation[i] === "-" || equation[i] === "+") {
            if (equation[i] === "-") {
                curValue = hasLeft ? "-" : "";
            }
            i++;
        }

        while (i < n && !isNaN(parseInt(equation[i]))) {
            curValue += equation[i];
            i++;
        }

        if (equation[i] === "x") {
            if (curValue === "-") {
                curValue = "-1";
            }
            if (curValue === "") {
                curValue = "1";
            }
            coefficient += Number(curValue);
            i++;
        } else {
            constant += Number(curValue);
        }
    }

    if (coefficient === 0 && constant === 0) {
        return "Infinite solutions";
    }

    constant = -constant;
    constant /= coefficient;

    // 判断结果是否为整数
    if (coefficient === 0 || constant % 1 !== 0) {
        return "No solution";
    }

    return "x=" + constant;
}

File: src/02.string/04.transform/08.640.go
package transform

import (
	"strconv"
	"unicode"
)

func SolveEquation(equation string) string {
	coefficient, constant := 0, 0
	i, n, hasLeft := 0, len(equation), true

	for i < n {
		if equation[i] == '=' {
			i++
			hasLeft = false
			continue
		}

		v := "-"
		if hasLeft {
			v = ""
		}
		if equation[i] == '-' || equation[i] == '+' {
			if equation[i] == '-' {
				v = ""
				if hasLeft {
					v = "-"
				}
			}
			i++
		}

		for i < n && unicode.IsDigit(rune(equation[i])) {
			v += string(equation[i])
			i++
		}

		if i < n && equation[i] == 'x' {
			if v == "" {
				v = "1"
			} else if v == "-" {
				v = "-1"
			}
			num1, _ := strconv.Atoi(v)
			coefficient += num1
			i++
		} else {
			num2, _ := strconv.Atoi(v)
			constant += num2
		}
	}

	if coefficient == 0 && constant == 0 {
		return "Infinite solutions"
	}

	if coefficient == 0 {
		return "No solution"
	}

	res := -constant / coefficient

	// x % 1 is always zero
	// if (res % 1) != 0 {
	// 	return "No solution"
	// }

	return "x=" + strconv.Itoa(res)
}
